p2p/switch.go

```diff
@@ -551,131 +519,82 @@ func (sw *Switch) addOutboundPeerWithConfig(
 	persistent bool,
 ) error {
 	sw.Logger.Info("Dialing peer", "address", addr)
-	peerConn, err := newOutboundPeerConn(
-		addr,
-		config,
-		persistent,
-		sw.nodeKey.PrivKey,
-	)
+
+	p, err := sw.transport.Dial(*addr, peerConfig{
+		chDescs:      sw.chDescs,
+		onPeerError:  sw.StopPeerForError,
+		persistent:   persistent,
+		reactorsByCh: sw.reactorsByCh,
+	})
 	if err != nil {
+		switch e := err.(type) {
+		case *ErrRejected:
+			if e.IsSelf() {
+				// Remove the given address from the address book and add to our addresses
+				// to avoid dialing in the future.
+				sw.addrBook.RemoveAddress(addr)
+				sw.addrBook.AddOurAddress(addr)
+			}
+		}
+
 		if persistent {
 			go sw.reconnectToPeer(addr)
 		}
-		return err
-	}
 
-	if err := sw.addPeer(peerConn); err != nil {
-		peerConn.CloseConn()
 		return err
 	}
-	return nil
-}
-
-// addPeer performs the Tendermint P2P handshake with a peer
-// that already has a SecretConnection. If all goes well,
-// it starts the peer and adds it to the switch.
-// NOTE: This performs a blocking handshake before the peer is added.
-// NOTE: If error is returned, caller is responsible for calling
-// peer.CloseConn()
-func (sw *Switch) addPeer(pc peerConn) error {
 
-	addr := pc.conn.RemoteAddr()
-	if err := sw.FilterConnByAddr(addr); err != nil {
+	if err := sw.addPeer(p); err != nil {
+		_ = p.Stop()
 		return err
 	}
 
-	// Exchange NodeInfo on the conn
-	peerNodeInfo, err := pc.HandshakeTimeout(sw.nodeInfo, time.Duration(sw.config.HandshakeTimeout))
-	if err != nil {
-		return err
-	}
-
-	peerID := peerNodeInfo.ID
-
-	// ensure connection key matches self reported key
-	connID := pc.ID()
-
-	if peerID != connID {
-		return fmt.Errorf(
-			"nodeInfo.ID() (%v) doesn't match conn.ID() (%v)",
-			peerID,
-			connID,
-		)
-	}
-
-	// Validate the peers nodeInfo
-	if err := peerNodeInfo.Validate(); err != nil {
-		return err
-	}
-
-	// Avoid self
-	if sw.nodeKey.ID() == peerID {
-		addr := peerNodeInfo.NetAddress()
-		// remove the given address from the address book
-		// and add to our addresses to avoid dialing again
-		if sw.addrBook != nil {
-			sw.addrBook.RemoveAddress(addr)
-			sw.addrBook.AddOurAddress(addr)
-		}
-		return ErrSwitchConnectToSelf{addr}
-	}
+	return nil
+}
 
+// addPeer starts up the Peer and adds it to the Switch.
+func (sw *Switch) addPeer(p Peer) error {
 	// Avoid duplicate
-	if sw.peers.Has(peerID) {
-		return ErrSwitchDuplicatePeerID{peerID}
+	if sw.peers.Has(p.ID()) {
+		return ErrSwitchDuplicatePeerID{p.ID()}
 	}
 
-	// Check for duplicate connection or peer info IP.
-	if !sw.config.AllowDuplicateIP &&
```

Comment:
noting that we may want to keep the ability to allow duplicate ips, eg. for local testing scenarios (though this probably better solved through a "dev" mode for tendermint than this specific config option ...) 

Also noting/asking we probably want to enforce the duplicate IP check on both the connection IP and the nodeInfo.IP

Reply:
Those concerns have been moved to the Transport filters. We allow for duplicate IPs by default unless the filter has been passed as option to the transport.

Wit the addition of filters on transport we have now two layers one on raw network connection and one on incoming peers. The check for the nodeinfo.IP sounds like should be best handled as a peer filter.